homeassistant/components/tado/coordinator.py

```diff
@@ -0,0 +1,390 @@
+            )
+        except PyTado.exceptions.TadoWrongCredentialsException as err:
+            raise ConfigEntryError(f"Invalid Tado credentials. Error: {err}") from err
+        except PyTado.exceptions.TadoException as err:
+            raise UpdateFailed(f"Error during Tado setup: {err}") from err
+        _LOGGER.debug("Tado connection established for username: %s", self._username)
+
+        try:
+            tado_home_call = await self.hass.async_add_executor_job(self.tado.get_me)
+            tado_home = tado_home_call["homes"][0]
+            self.home_id = tado_home["id"]
+            self.home_name = tado_home["name"]
+
+            _LOGGER.debug("Preloading zones and devices")
+            self.zones = await self.hass.async_add_executor_job(self.tado.get_zones)
+            self.devices = await self.hass.async_add_executor_job(self.tado.get_devices)
+        except Exception as err:
+            raise UpdateFailed(f"Error during Tado setup: {err}") from err
+        _LOGGER.info("Tado setup complete")
+
+    async def _async_update_data(self) -> dict[str, dict]:
+        """Fetch the latest data from Tado."""
+
+        try:
+            # Fetch updated data for devices, mobile devices, zones, and home
+            await self._async_update_devices()
+            await self._async_update_mobile_devices()
+            await self._async_update_zones()
+            await self._async_update_home()
+        except Exception as err:
+            raise UpdateFailed(f"Error updating Tado data: {err}") from err
+
+        return self.data
+
+    async def _async_update_devices(self) -> None:
+        """Update the device data from Tado."""
+        if self.tado is None:
+            raise UpdateFailed("Tado client is not initialized")
+
+        try:
+            devices = await self.hass.async_add_executor_job(self.tado.get_devices)
+        except RequestException as err:
+            _LOGGER.error("Error updating Tado devices: %s", err)
+            raise UpdateFailed(f"Error updating Tado devices: {err}") from err
+
+        if not devices:
+            _LOGGER.error("No linked devices found for home ID %s", self.home_id)
+            raise UpdateFailed(f"No linked devices found for home ID {self.home_id}")
+
+        for device in devices:
+            device_short_serial_no = device["shortSerialNo"]
+            _LOGGER.debug("Updating device %s", device_short_serial_no)
+            try:
+                if (
+                    INSIDE_TEMPERATURE_MEASUREMENT
+                    in device["characteristics"]["capabilities"]
+                ):
+                    _LOGGER.debug(
+                        "Updating temperature offset for device %s",
+                        device_short_serial_no,
+                    )
+                    device[TEMP_OFFSET] = await self.hass.async_add_executor_job(
+                        self.tado.get_device_info, device_short_serial_no, TEMP_OFFSET
+                    )
```

Comment:
Switching from the event loop to the executor job is quite expensive, so we should avoid creating too many executor jobs for similar actions. So let's try to put the whole `for` loop in the executor rather than creating executor jobs in the loop